sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1264,17 +1668,21 @@ public Object restriction() {
             .setTransformId(pTransformId)
             .setInputId(mainInputId)
             .setElement(residualBytes.toByteString());
-
+    Map<String, org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.Timestamp>
+        outputWatermarkMap = new HashMap<>();
     if (!watermarkAndState.getKey().equals(GlobalWindow.TIMESTAMP_MIN_VALUE)) {
+      org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.Timestamp outputWatermark =
+          org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.Timestamp.newBuilder()
+              .setSeconds(watermarkAndState.getKey().getMillis() / 1000)
+              .setNanos((int) (watermarkAndState.getKey().getMillis() % 1000) * 1000000)
+              .build();
       for (String outputId : pTransform.getOutputsMap().keySet()) {
-        residualApplication.putOutputWatermarks(
-            outputId,
-            org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.Timestamp.newBuilder()
-                .setSeconds(watermarkAndState.getKey().getMillis() / 1000)
-                .setNanos((int) (watermarkAndState.getKey().getMillis() % 1000) * 1000000)
-                .build());
+        outputWatermarkMap.put(outputId, outputWatermark);
       }
     }
+    if (!outputWatermarkMap.isEmpty()) {
+      residualApplication.putAllOutputWatermarks(outputWatermarkMap);
```

Comment:
Ditto, population the output map with an empty input map should be a no-op making the `if` check extraneous.